### PR TITLE
Properly escape local paths into URLs in fetchTree

### LINKED ISSUE #7707: `builtins.fetchGit ./.` fails when current working dir local path has spaces.
**Describe the bug**

`builtins.fetchGit` fails when used with a local repository the filepath of which contains spaces.

**Steps To Reproduce**

1. Create a `git` repository in a directory where either that directory or a parent directory has spaces in its filepath.
2. Navigate to that repository.
3. Try `builtins.fetchGit ./.` as below:

```Console
$ nix repl
Welcome to Nix 2.12.0. Type :? for help.

nix-repl> builtins.fetchGit ./.
error: 'file:///home/User/Files/My Project Files/master' is not a valid URL
```

**Expected behavior**

`builtins.fetchGit` should handle spaces in local filepaths.

**Additional context**

`pkgs.fetchgit` does not have this problem.

### PR DESCRIPTION
**Pull request by yorickvP**
# Motivation
`fetchTree` was converting local paths into urls using ("file://" + path), resulting in errors when paths contained url-illegal characters, such as spaces.

# Context
Fixes #7707

- Remove all reserved characters from `percentEncode` (encoding too much isn't a problem).
- Add unit tests
- Use it for encoding URLs
- Use `ParsedURL::to_string` to convert paths to URLs in fetchTree

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [x] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or bug fix: updated release notes

### CODE REVIEW

File: src/libexpr/primops/fetchTree.cc [hunk 1]
```diff
@@ -68,7 +69,16 @@ void emitTreeAttrs(
 std::string fixURI(std::string uri, EvalState & state, const std::string & defaultScheme = "file")
```
Comment (roberth) [inline src/libexpr/primops/fetchTree.cc:69]: Yikes. What is this about? URIs can't be fixed.

I wish we'd just use (semi-)structured data. URIs are such a ~mess~ burden.

Comment (yorickvP) [reply src/libexpr/primops/fetchTree.cc:69]: This is the code that changes passed paths and strings to `fetchTree` to be an URL.

Examples: `"git@asdf:foo.git"`, `./.`, `"https://github.com/nixos/nixpkgs`

Sadly, `fetchers::Input::fromAttrs` expects to receive a standard attrset with an url-formatted string, so this goes through `ParsedURL` twice.

Comment (roberth) [reply src/libexpr/primops/fetchTree.cc:69]: > a standard attrset with an url-formatted string

That should never have been a thing. When working with any kind of serialization, including URIs, we must keep it simple by only having our data in one of two states: serialized, or deserialized. Doing string operations on URIs is asking for trouble.

**Review by roberth — APPROVED** [review]
This looks like the right thing: escaping when it's encoding a URL. There should be no reason to put an escaped value in a decoded field (ie in a struct). I'm not too familiar with the fetchTree code yet, so I'll defer the decision to merge to @edolstra.

**Comment by thufschmitt** [comment]
Is this interacting with https://github.com/NixOS/nix/pull/6614/ ? At a glance it seems like they are complementary but I'd like a confirmation

File: src/libutil/url.cc [hunk 2]
```diff
@@ -88,17 +88,22 @@ std::map<std::string, std::string> decodeQuery(const std::string & query)
...
 {
     std::string res;
     for (auto & c : s)
+        // unreserved + keep
         if ((c >= 'a' && c <= 'z')
             || (c >= 'A' && c <= 'Z')
             || (c >= '0' && c <= '9')
-            || strchr("-._~!$&'()*+,;=:@", c))
+            || strchr("-._~", c)
+            || keep.find(c) != std::string::npos)
```
Comment (thufschmitt) [inline src/libutil/url.cc:103]: Why do we need to make this stricter? Was the old behaviour incorrect?

**Review by thufschmitt — CHANGES_REQUESTED** [review]
Looks good as far as the implementation is concerned :+1: 

Can you add a regression test for https://github.com/NixOS/nix/issues/7707 with it? So that it stays properly fixed

File: src/libutil/url.cc (hunk 2 above)
Comment (yorickvP) [reply src/libutil/url.cc:103]: The removed characters belong to a "reserved" set, which have a function in some parts of the URL (such as `&` being a query param separator).

It is always safe to escape characters even if they are allowed, but `keep` has been added for when you know which characters are allowed in the part of the URL you are encoding.

**Pull request merged**